File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/ShareUtilsImpl.java
package com.garpr.android.misc;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ShareCompat;
import android.text.TextUtils;
import android.widget.Toast;

import com.garpr.android.R;
import com.garpr.android.models.AbsPlayer;
import com.garpr.android.models.AbsTournament;
import com.garpr.android.models.Region;

import static com.garpr.android.misc.Constants.PLAIN_TEXT;

public class ShareUtilsImpl implements ShareUtils {

    private static final String TAG = "ShareUtilsImpl";

    private final RegionManager mRegionManager;
    private final Timber mTimber;


    public ShareUtilsImpl(@NonNull final RegionManager regionManager,
            @NonNull final Timber timber) {
        mRegionManager = regionManager;
        mTimber = timber;
    }

    @Override
    public void openUrl(@NonNull final Context context, @Nullable final String url) {
        if (TextUtils.isEmpty(url) || TextUtils.getTrimmedLength(url) == 0) {
            return;
        }

        final Uri uri = Uri.parse(url);

        try {
            final Intent intent = new Intent(Intent.ACTION_VIEW, uri);
            context.startActivity(intent);
        } catch (final ActivityNotFoundException e) {
            mTimber.e(TAG, "Unable to open browser to URI: \"" + uri + "\"", e);
            Toast.makeText(context, R.string.unable_to_open_link, Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public void sharePlayer(@NonNull final Activity activity, @NonNull final AbsPlayer player) {
        final Region region = mRegionManager.getRegion(activity);

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_x, player.getName()))
                .setText(region.getEndpoint().getPlayerWebPath(region.getId(), player.getId()))
                .setType(PLAIN_TEXT)
                .startChooser();
    }

    @Override
    public void shareRankings(@NonNull final Activity activity) {
        final Region region = mRegionManager.getRegion(activity);

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_rankings))
                .setText(region.getEndpoint().getRankingsWebPath(region.getId()))
                .setType(PLAIN_TEXT)
                .startChooser();
    }

    @Override
    public void shareTournament(@NonNull final Activity activity,
            @NonNull final AbsTournament tournament) {
        final Region region = mRegionManager.getRegion(activity);

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_x, tournament.getName()))
                .setText(region.getEndpoint().getTournamentWebPath(region.getId(), tournament.getId()))
                .setType(PLAIN_TEXT)
                .startChooser();
    }

    @Override
    public void shareTournaments(@NonNull final Activity activity) {
        final Region region = mRegionManager.getRegion(activity);

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_tournaments))
                .setText(region.getEndpoint().getTournamentsWebPath(region.getId()))
                .setType(PLAIN_TEXT)
                .startChooser();
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/ShareUtilsImpl.kt
package com.garpr.android.misc

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.widget.Toast
import androidx.core.app.ShareCompat
import com.garpr.android.R
import com.garpr.android.data.models.AbsPlayer
import com.garpr.android.data.models.AbsTournament
import com.garpr.android.misc.Constants.PLAIN_TEXT
import android.net.Uri as AndroidUri

class ShareUtilsImpl(
        private val regionHandleUtils: RegionHandleUtils,
        private val timber: Timber
) : ShareUtils {

    override fun openUrl(context: Context, url: String?) {
        if (url.isNullOrBlank()) {
            return
        }

        val androidUri = AndroidUri.parse(url)
        val intent = Intent(Intent.ACTION_VIEW, androidUri)

        try {
            context.startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            timber.e(TAG, "Unable to open browser to URI: $androidUri", e)
            Toast.makeText(context, R.string.unable_to_open_link, Toast.LENGTH_LONG).show()
        }
    }

    override fun sharePlayer(activity: Activity, player: AbsPlayer) {
        val region = regionHandleUtils.getRegion(activity)

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_x, player.name))
                .setText(region.endpoint.getPlayerWebPath(region.id, player.id))
                .setType(PLAIN_TEXT)
                .startChooser()
    }

    override fun shareRankings(activity: Activity) {
        val region = regionHandleUtils.getRegion(activity)

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_rankings))
                .setText(region.endpoint.getRankingsWebPath(region.id))
                .setType(PLAIN_TEXT)
                .startChooser()
    }

    override fun shareTournament(activity: Activity, tournament: AbsTournament) {
        val region = regionHandleUtils.getRegion(activity)

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_x, tournament.name))
                .setText(region.endpoint.getTournamentWebPath(region.id, tournament.id))
                .setType(PLAIN_TEXT)
                .startChooser()
    }

    override fun shareTournaments(activity: Activity) {
        val region = regionHandleUtils.getRegion(activity)

        ShareCompat.IntentBuilder.from(activity)
                .setChooserTitle(activity.getString(R.string.share_tournaments))
                .setText(region.endpoint.getTournamentsWebPath(region.id))
                .setType(PLAIN_TEXT)
                .startChooser()
    }

    companion object {
        private const val TAG = "ShareUtilsImpl"
    }

}
